clang/lib/Sema/SemaDecl.cpp

```diff
@@ -15811,6 +15813,20 @@ static void diagnoseImplicitlyRetainedSelf(Sema &S) {
           << FixItHint::CreateInsertion(P.first, "self->");
 }
 
+void Sema::CheckCoroutineWrapper(FunctionDecl *FD) {
+  if (!FD || getCurFunction()->isCoroutine())
```

Comment:
Why are we disabling this analysis if the current function is a coroutine?
It seems like a red flag, i.e. does this mean that a lambda inside a coroutine can violate the contract? What's the motivation there.

Or am I missing something?

Reply:
`getCurFunction()` returns the function scope info which contains information about the processed function body being a coroutine or not. FD and getCurFunction() refers to the same function (one is decl and the other is FSI).

I will add a lambda inside a coroutine test to clarify.

Reply:
It's not clear at all from the code now that `FD` and `getCurFunction()` must correspond to the code.
As we chatted offline, moving the check `getCurFunction()->isCoroutine()` into the callers and checking only the attributes in this function would make it much easier to understand the code.

Let's do that?

Alternatively, we could add some assertion that checks the `getCurFunction()` and `FD` correspond to the same function.

Reply:
Moved `getCurFunction()` out and updated docs to say that it expects FD to be non-coroutine.